pkg/controller/snapshot_controller.go

```diff
@@ -0,0 +1,856 @@
+func (ctrl *csiSnapshotController) createSnapshot(snapshot *crdv1.VolumeSnapshot) error {
+	glog.V(5).Infof("createSnapshot[%s]: started", snapshotKey(snapshot))
+	opName := fmt.Sprintf("create-%s[%s]", snapshotKey(snapshot), string(snapshot.UID))
+	ctrl.scheduleOperation(opName, func() error {
+		snapshotObj, err := ctrl.createSnapshotOperation(snapshot)
+		if err != nil {
+			ctrl.updateSnapshotErrorStatusWithEvent(snapshot, v1.EventTypeWarning, "SnapshotCreationFailed", fmt.Sprintf("Failed to create snapshot: %v", err))
+			glog.Errorf("createSnapshot [%s]: error occurred in createSnapshotOperation: %v", opName, err)
+			return err
+		}
+		_, updateErr := ctrl.storeSnapshotUpdate(snapshotObj)
+		if updateErr != nil {
+			// We will get an "snapshot update" event soon, this is not a big error
+			glog.V(4).Infof("createSnapshot [%s]: cannot update internal cache: %v", snapshotKey(snapshotObj), updateErr)
+		}
+
+		return nil
+	})
+	return nil
+}
+
+func (ctrl *csiSnapshotController) checkandUpdateSnapshotStatus(snapshot *crdv1.VolumeSnapshot, content *crdv1.VolumeSnapshotContent) error {
+	glog.V(5).Infof("checkandUpdateSnapshotStatus[%s] started", snapshotKey(snapshot))
+	opName := fmt.Sprintf("check-%s[%s]", snapshotKey(snapshot), string(snapshot.UID))
+	ctrl.scheduleOperation(opName, func() error {
+		snapshotObj, err := ctrl.checkandUpdateSnapshotStatusOperation(snapshot, content)
+		if err != nil {
+			ctrl.updateSnapshotErrorStatusWithEvent(snapshot, v1.EventTypeWarning, "SnapshotCheckandUpdateFailed", fmt.Sprintf("Failed to check and update snapshot: %v", err))
+			glog.Errorf("checkandUpdateSnapshotStatus [%s]: error occured %v", snapshotKey(snapshot), err)
+			return err
+		}
+		_, updateErr := ctrl.storeSnapshotUpdate(snapshotObj)
+		if updateErr != nil {
+			// We will get an "snapshot update" event soon, this is not a big error
+			glog.V(4).Infof("checkandUpdateSnapshotStatus [%s]: cannot update internal cache: %v", snapshotKey(snapshotObj), updateErr)
+		}
+
+		return nil
+	})
+	return nil
+}
+
+// updateSnapshotStatusWithEvent saves new snapshot.Status to API server and emits
+// given event on the snapshot. It saves the status and emits the event only when
+// the status has actually changed from the version saved in API server.
+// Parameters:
+//   snapshot - snapshot to update
+//   eventtype, reason, message - event to send, see EventRecorder.Event()
+func (ctrl *csiSnapshotController) updateSnapshotErrorStatusWithEvent(snapshot *crdv1.VolumeSnapshot, eventtype, reason, message string) error {
+	glog.V(5).Infof("updateSnapshotStatusWithEvent[%s]", snapshotKey(snapshot))
+
+	if snapshot.Status.Error != nil && snapshot.Status.Error.Message == message {
+		glog.V(4).Infof("updateSnapshotStatusWithEvent[%s]: the same error %v is already set", snapshot.Name, snapshot.Status.Error)
+		return nil
+	}
+	snapshotClone := snapshot.DeepCopy()
+	if snapshot.Status.Error == nil {
```

Comment:
remove this line to allow new discovered error to show up.

Reply:
Done